lib/services/resourceService.js

```diff
@@ -0,0 +1,284 @@
+
+    // ensure organisation records for practitioner are also fetched and cached
+    await P.each(practionerRoleBundle.entry, async (entry) => {
+
+      const role = entry.resource;
+
+      if (role.resourceType !== ResourceName.PRACTITIONERROLE) return;
+
+      const organisationRef = role.organization.reference;
+      const organisationResponse = await this.fetchResource(organisationRef);
+
+      if (organisationResponse.resource) {
+        resourceCache.byUuid.setRelatedUuid(ResourceName.PRACTITIONER, resource.id, ResourceName.ORGANIZATION, organisationResponse.resource.id);
+      }
+    });
+  }
+
+  /**
+   * Fetch a resource
+   *
+   * @param  {string} reference
+   * @return {Promise.<Object>}
+   */
+  async fetchResource(reference) {
+    logger.info('services/resourceService|fetchResource', { reference });
+
+    const { resourceName, uuid } = parseRef(reference);
+    const { fetchCache, resourceCache } = this.ctx.cache;
+
+    const exists  = resourceCache.byUuid.exists(resourceName, uuid);
+    if (exists) {
+      return {
+        ok: false,
+        exists: true
+      };
+    }
+
+    const fetching = fetchCache.exists(reference);
+    if (fetching) {
+      return {
+        ok: false,
+        fetching: true
+      };
+    }
+
+    const { tokenService, resourceRestService } = this.ctx.services;
+    const token = await tokenService.get();
+
+    fetchCache.set(reference);
+    const resource = await resourceRestService.getResource(reference, token);
+
+    debug('resource: %j', resource);
+
+    resourceCache.byUuid.set(resourceName, uuid, resource);
+
+    return {
+      ok: true,
+      resource
+    };
+  }
+
+  /**
+   * Fetch resources
+   *
+   * @param  {string} resourceType
+   * @param  {Object} queryParams
+   * @return {Promise.<Object>}
+   */
+  async fetchResources(resourceType, queryParams) {
+    logger.info('services/resourceService|fetchResources', { queryParams });
+
+    const query = this.mapQuery(resourceType, queryParams);
+
+    const { fetchCache, resourceCache } = this.ctx.cache;
+
+    const exists  = resourceCache.byQuery.exists(resourceType, query);
+    if (exists) {
+      return {
+        ok: false,
+        exists: true
+      };
+    }
+
+    const fetching = fetchCache.exists(query);
+    if (fetching) {
+      return {
+        ok: false,
+        fetching: true
+      };
+    }
+
+    const { tokenService, resourceRestService } = this.ctx.services;
+    const token = await tokenService.get();
+
+    fetchCache.set(query);
+    const resource = await resourceRestService.getResources(resourceType, query, token);
+
+    debug('resource: %j', resource);
+
+    resourceCache.byQuery.set(resourceType, query, resource);
+
+    return {
+      ok: true,
+      resource
+    };
+  }
+
+  /**
+   * Maps query parameters to FHIR search string
+   *
+   * @param  {Object} queryParams
+   * @return {string}
+   */
+  mapQuery(resourceType, queryParams) {
```

Comment:
@richardcbrown I'd recommend you to move `mapQuery` to `share/utils` because it doesn't depend on any context/this functionality.